io/directory.go

```diff
@@ -573,42 +591,21 @@
 //  sure we make good on the value). Finding the right margin can be tricky
 //  and very dependent on the use case so it might not be worth it.
 func (d *UpgradeableDirectory) RemoveChild(ctx context.Context, name string) error {
-	if err := d.Directory.RemoveChild(ctx, name); err != nil {
-		return err
-	}
-
 	hamtDir, ok := d.Directory.(*HAMTDirectory)
-	if !ok { // BasicDirectory
-		return nil
-	}
-
-	if HAMTShardingSize == 0 || // Option disabled.
-		hamtDir.sizeChange >= 0 { // We haven't reduced the HAMT net size.
-		return nil
-	}
-
-	// We have reduced the directory size, check if it didn't go under
-	// the HAMTShardingSize threshold.
-	belowThreshold, timeoutExceeded := hamtDir.sizeBelowThreshold(EvaluateHAMTTransitionTimeout)
-
-	if timeoutExceeded {
-		// We run out of time before confirming if we're indeed below the
-		// threshold. When in doubt error to not return inconsistent structures.
-		// FIXME: We could allow this to return without error and enforce this
-		//  timeout on a GetNode() call when we need to actually commit to a
-		//  structure/CID. (The downside is that GetNode() doesn't have a
-		//  context argument and we would have to break the API.)
-		return fmt.Errorf("not enought time to fetch shards")
-		// FIXME: Abstract in new error for testing.
-	}
-
-	if belowThreshold { // Switch.
-		basicDir, err := hamtDir.switchToBasic(ctx)
+	if ok {
```

Comment:
Good catch. The 'check decrease in size if overwrite' bug also applies to the basic directory, will fix.

Reply:
Actually the directionality of the switch makes this unnecessary (unless you have a specific scenario in mind). AFAIK a `(*BasicDirectory).RemoveChild()` call will _never_ increase the size of the directory and thus requiring the switch to a `HAMTDirectory`.

Still, the original comment flags that I need to fix `(*BasicDirectory).AddChild()` to account for a potential size decrease in case of overwriting (removing) an old entry.

Reply:
`(*BasicDirectory).addLinkChild()` (plumbing function for `(*BasicDirectory).AddChild()`) is already calling `(*BasicDirectory).RemoveChild()` internally which accounts for the reduction in size.

This whole thread then is a no-op; added a minor comment to make the size change more explicit.

Reply:
> Actually the directionality of the switch makes this unnecessary (unless you have a specific scenario in mind).

> While this will almost never be the case the check should be really cheap and will help us out if people call RemoveChild on directories that are already too big.

A scenario is that folks who have BasicDirectories that are already too big (e.g. for legacy reasons), but that we can still read will switch into our new realm of sharded directories any time the do a modification. We can't do much for people who don't modify their directories, but as soon as they make any changes (and therefore would end up with a new CID) we should try to make things work for them.